VictoryCursorContainer

chrisbolin opened the pull request:
Implements VictoryCursorContainer (https://github.com/FormidableLabs/victory/issues/514)

demo at http://localhost:3000/#/cursor-container

Open issues:
- Where should the label go for 1-dimensional cursors? (with mouse [current implementation], outside of chart, middle of chart, something else?)
- createContainer("cursor", "voronoi") works pretty well, but is awkward when `labels` is used (both containers try to create labels, but they are in different places and for different values)

chrisbolin commented:
### VictoryCursorContainer

![may-11-2017 19-59-38](https://cloud.githubusercontent.com/assets/2822048/25964240/7c1a8e5a-3684-11e7-8172-37bd27e5cdd6.gif)

chrisbolin commented:
### `createContainer("cursor", "voronoi")`

![may-11-2017 19-59-06](https://cloud.githubusercontent.com/assets/2822048/25964351/d7b96344-3684-11e7-8c93-3dfae63611c5.gif)

boygirl commented on src/components/containers/victory-cursor-container.js:
```diff
@@ -0,0 +1,123 @@
+  static displayName = "VictoryCursorContainer";
+  static propTypes = {
+    ...VictoryContainer.propTypes,
+    defaultCursorValue: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
```
Prefer `PropTypes.shape(x... y...)` to `PropTypes.object`

chrisbolin replied:
yup!

boygirl commented on src/components/containers/victory-cursor-container.js:
```diff
@@ -0,0 +1,123 @@
+    return newElements;
+  }
+
+  getTooltip(props) {
```
`getLabel`

boygirl commented on src/components/containers/victory-cursor-container.js:
```diff
@@ -0,0 +1,123 @@
+        y: cursorCoordinates.y - 10,
+        text: Helpers.evaluateProp(labels, cursorPosition, true),
+        active: true,
+        key: "cursor-tooltip"
```
"cursor-label"

boygirl commented on src/components/containers/victory-cursor-container.js:
```diff
@@ -0,0 +1,123 @@
+    };
+    if (labels) {
+      newElements.push(React.cloneElement(labelComponent, {
+        x: cursorCoordinates.x + 5,
```
hard coded positioning here seems wrong. Maybe a `labelOffset` prop that can take a single value, or values for x and y.

boygirl replied:
The values you have here can be the defaults

chrisbolin replied:
thank you! i felt the same way

boygirl commented on src/components/containers/victory-cursor-container.js:
```diff
@@ -0,0 +1,123 @@
+    const newElements = [];
+    const domainCoordinates = Selection.getDomainCoordinates(scale, domain);
+
+    const cursorCoordinates = {
```
You're already getting the svg coordinates of the mouse position and transforming those into data coordinates in your helpers. You could just add the svg coordinates to the mutation in `onMouseMove`

chrisbolin replied:
totally understand. i think i'll keep it this way though, because it handles getting the position when `defaultCursorValue` (which can be both number and object) is used (i.e. there is no `onMouseMove`).

boygirl replied:
👍

boygirl commented:
Oh! actually, I just noticed... do `labels` and `labelComponent` conflict when you use `createContainer` with "voronoi" and "cursor"? How would you define independent values?

chrisbolin commented:
@boygirl YES, you are correct. When `createContainer` is used, they will both reference the same `labels` prop. That's why the example above (top-left under `createContainer("cursor", "voronoi")`) has two tooltips.

We could rename the props. This is definitely the simplest solution, but also is kind of skirting the problem. But hey, being specific (like with `voronoiLabels` and `cursorLabelComponent`) does make things more clear.

I'm not sure exactly what another solution would look like. `createContainer` literally creates a single class that implements the hybrid container.

boygirl commented:
In that case, let's go with `cursorLabels` and `cursorLabelComponent`. We can leave `VictoryVoronoiContainer` as is to avoid a breaking change

chrisbolin commented:
@boygirl addressed those issues. waiting for CI to pass before I do anything.

also, i went with `cursorLabel` (singular) instead of `cursorLabels`, as a cursor can only have one label, and cursorLabelComponent was singular. Totally happy to do plural `cursorLabels` if you think that would be more consistent with other components.

boygirl commented:
@chrisbolin looks good to me! Approved!

boygirl commented:
I'm about to do a quick release, so I'm going to go ahead and merge this :)
